source/mesh_refinement/isosurfaces.cc

```diff
@@ -0,0 +1,391 @@
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+
+#include <aspect/mesh_refinement/isosurfaces.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+#include <math.h>
+#include <algorithm>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    namespace Internal
+    {
+      bool
+      Isosurface::values_are_in_range(const std::vector<double> values) const
+      {
+        // This assumes that all the vectors in isosurfaces already have the
+        // same length.
+        Assert(values.size() == min_values.size(),
+               ExcMessage("Internal error: Vector of values passed to the isosurface class function values_are_in_range, "
+                          "does not have the the correct size."));
+        for (unsigned int index = 0; index < values.size(); ++index)
+          {
+            if (values[index] < min_values[index] || values[index] > max_values[index])
+              {
+                // outside this isosurface, no need to search further
+                return false;
+              }
+          }
+        // If we made it this far, then we are inside the conditions, so return true.
+        return true;
+      }
+
+      Property::Property(const std::string &property_name,const std::vector<std::string> &available_compositions)
+      {
+        bool found = false;
+        if (property_name == "Temperature")
+          {
+            name = PropertyName::Temperature;
+            index = 0;
+            found = true;
+          }
+        else if (property_name == "background")
+          {
+            name = PropertyName::Background;
+            index = 0;
+            found = true;
+          }
+        else
+          {
+            auto p = std::find(available_compositions.begin(), available_compositions.end(), property_name);
+            if (p != available_compositions.end())
+              {
+                name = PropertyName::Composition;
+                index = std::distance(available_compositions.begin(), p);
+                found = true;
+              }
+          }
+        if (found == false)
+          {
+            // The property name has not been found. This could come from that the compositional field was not present.
+            // Abort and warn the user.
+            std::string key_list = "Temperature";
+            for (auto &composition : available_compositions)
+              key_list += ", " + composition;
+
+            AssertThrow(false,
+                        ExcMessage("The key given for the isosurface could not be converted. The provided key was: " + property_name + ". "
+                                   "The following keys are allowed for this model: " + key_list + "."));
+          }
+      }
+
+      unsigned int min_max_string_to_int(const std::string &string_value, const unsigned int minimum_refinement_level, const unsigned int  maximum_refinement_level)
```

Comment:
I don't understand what this function does. You seem to be parsing an expression here. Consider using a FunctionParser or can you write a few unittests?

Reply:
I have moved it, documented it and added a unit test :)